felix/bpf/nat/connecttime.go

```diff
@@ -161,60 +204,93 @@ func InstallConnectTimeLoadBalancer(ipFamily int, cgroupv2 string, logLevel stri
 		return errors.Wrap(err, "failed to set-up cgroupv2")
 	}
 
-	switch ipFamily {
-	case 4:
-		err = installProgram("connect", "4", bpfMount, cgroupPath, logLevel, udpNotSeen, excludeUDP)
+	ctlbProgsMap := NewProgramsMap()
+	var v4Obj, v46Obj, v6Obj *libbpf.Obj
+
+	// Load the v6 CTLB program.
+	if ipv6Enabled {
+		v6Obj, err = loadProgram(logLevel, "6", udpNotSeen, excludeUDP)
+		if err != nil {
+			return err
+		}
+		defer v6Obj.Close()
+	}
+
+	// Load and attach v4, v46 CTLB program.
+	if ipv4Enabled {
+		v4Obj, err = loadProgram(logLevel, "4", udpNotSeen, excludeUDP)
 		if err != nil {
 			return err
 		}
+		defer v4Obj.Close()
 
-		err = installProgram("connect", "46", bpfMount, cgroupPath, logLevel, udpNotSeen, excludeUDP)
+		v46Obj, err = loadProgram(logLevel, "46", udpNotSeen, excludeUDP)
+		if err != nil {
+			return err
+		}
+		defer v46Obj.Close()
+		err = attachProgram("connect", "4", bpfMount, cgroupPath, udpNotSeen, excludeUDP, v4Obj)
+		if err != nil {
+			return err
+		}
+		err = attachProgram("connect", "46", bpfMount, cgroupPath, udpNotSeen, excludeUDP, v46Obj)
 		if err != nil {
 			return err
 		}
 
 		if !excludeUDP {
-			err = installProgram("sendmsg", "4", bpfMount, cgroupPath, logLevel, udpNotSeen, false)
+			err = attachProgram("sendmsg", "4", bpfMount, cgroupPath, udpNotSeen, false, v4Obj)
 			if err != nil {
 				return err
 			}
 
-			err = installProgram("recvmsg", "4", bpfMount, cgroupPath, logLevel, udpNotSeen, false)
+			err = attachProgram("recvmsg", "4", bpfMount, cgroupPath, udpNotSeen, false, v4Obj)
 			if err != nil {
 				return err
 			}
 
-			err = installProgram("sendmsg", "46", bpfMount, cgroupPath, logLevel, udpNotSeen, false)
+			err = attachProgram("sendmsg", "46", bpfMount, cgroupPath, udpNotSeen, false, v46Obj)
 			if err != nil {
 				return err
 			}
 
-			err = installProgram("recvmsg", "46", bpfMount, cgroupPath, logLevel, udpNotSeen, false)
+			err = attachProgram("recvmsg", "46", bpfMount, cgroupPath, udpNotSeen, false, v46Obj)
 			if err != nil {
 				return err
 			}
 		}
-	case 6:
-		err = installProgram("connect", "6", bpfMount, cgroupPath, logLevel, udpNotSeen, excludeUDP)
-		if err != nil {
-			return err
-		}
-
-		if !excludeUDP {
-			err = installProgram("sendmsg", "6", bpfMount, cgroupPath, logLevel, udpNotSeen, false)
-			if err != nil {
+		// If dual-stack, populate the jump maps with v6 ctlb programs.
+		if v6Obj != nil {
+			if err := ctlbProgsMap.EnsureExists(); err != nil {
+				log.WithError(err).Error("Failed to create CTLB programs maps")
 				return err
 			}
-
-			err = installProgram("recvmsg", "6", bpfMount, cgroupPath, logLevel, udpNotSeen, false)
+			err = updateCTLBJumpMap(ctlbProgsMap.GetName(), v6Obj)
```

Comment:
My expectation would be that you update the map in v46Obj. It just happens that you have the map in all of the 3 objects and pinned to the same one, so this update is "correct", but if you had it in v46 only, this would not work imo. v6 is not going to jump anywhere.